### PR TITLE
VJP/JVP support pytree

### PR DESCRIPTION
**Pull request by rmoyard**
**Context:**

Following https://github.com/PennyLaneAI/catalyst/pull/500, we aim to add support for arbitrary return of functions for VJP and JVP.

**Description of the Change:**

- JVP and VJP are updated to support pytree as return.
- Clean the tests.

### CODE REVIEW

**Comment by rmoyard** [comment]
[sc-55113]

**Review by erick-xanadu — APPROVED** [review]
I am not too familiar with JVPs not VJPs, so some more comments would be nice. But the code looks good! I.e., why is there a midpoint in one of the options, and why is the shape of the VJP the same as the parameters?

File: frontend/catalyst/pennylane_extensions.py [hunk 1]
```diff
@@ -935,9 +945,21 @@ def check_is_iterable(x, hint):
...
+            func_res = results[: len(jaxpr.out_avals)]
+            vjps = results[len(jaxpr.out_avals) :]
+            func_res = tree_unflatten(out_tree, func_res)
+            # We do not change the shape of the VJP as it is the same as the parameters not
+            # as the output
+            results = tuple([func_res, tuple(vjps)])
+        else:
+            func_res = results[: len(jaxpr.out_avals)]
+            vjps = results[len(jaxpr.out_avals) :]
+            results = tuple([*func_res, tuple(vjps)])
```
Comment (dime10) [inline frontend/catalyst/pennylane_extensions.py:962]: This structure seems a bit strange no? The function results are expanded but the vjps are in another tuple

**Comment by rmoyard** [comment]
@erick-xanadu It is because the JVP have the same shape as the returns, where VJP have the same shape as the parameters.

File: frontend/catalyst/pennylane_extensions.py [hunk 2]
```diff
@@ -935,9 +945,21 @@ def check_is_iterable(x, hint):
...
 
         results = vjp_p.bind(*params, *cotangents, jaxpr=jaxpr, fn=fn, grad_params=grad_params)
+
+        if out_tree.children() != []:
+            func_res = results[: len(jaxpr.out_avals)]
+            vjps = results[len(jaxpr.out_avals) :]
+            func_res = tree_unflatten(out_tree, func_res)
+            # We do not change the shape of the VJP as it is the same as the parameters not
+            # as the output
+            results = tuple([func_res, tuple(vjps)])
```
Comment (dime10) [inline frontend/catalyst/pennylane_extensions.py:958]: I was considering if the vjps should be unflattened with the function argument trees, but probably not? 😅

**Review by dime10 — COMMENTED** [review]
How does the behaviour match JAX, is it 1-1 or are there certain deviations?

File: frontend/catalyst/pennylane_extensions.py [hunk 3]
```diff
@@ -935,9 +945,21 @@ def check_is_iterable(x, hint):
         fn = _ensure_differentiable(f)
         grad_params = _check_grad_params(method, scalar_out, h, argnum)
 
-        jaxpr, _ = _make_jaxpr_check_differentiable(fn, grad_params, *params)
+        jaxpr, out_tree = _make_jaxpr_check_differentiable(fn, grad_params, *params)
 
         results = vjp_p.bind(*params, *cotangents, jaxpr=jaxpr, fn=fn, grad_params=grad_params)
```
Comment (dime10) [inline frontend/catalyst/pennylane_extensions.py:950]: Shouldn't we flatten the params and cotangents before binding?

File: frontend/catalyst/pennylane_extensions.py (hunk 2 above)
Comment (dime10) [reply frontend/catalyst/pennylane_extensions.py:958]: Well the cotangent arguments actually

File: frontend/catalyst/pennylane_extensions.py [hunk 4]
```diff
@@ -869,9 +869,19 @@ def check_is_iterable(x, hint):
...
+
+        if out_tree.children() != []:
+            midpoint = len(results) // 2
+            func_res = results[:midpoint]
+            jvps = results[midpoint:]
+            func_res = tree_unflatten(out_tree, func_res)
+            jvps = tree_unflatten(out_tree, jvps)
+            results = tuple([func_res, jvps])
+        else:
+            results = tuple(results)
```
Comment (dime10) [inline frontend/catalyst/pennylane_extensions.py:884]: Is this not inconsistent? I.e. in one branch the results and jvps are flat, but in the other they are separated into a pair?

File: frontend/catalyst/pennylane_extensions.py (hunk 3 above)
Comment (rmoyard) [reply frontend/catalyst/pennylane_extensions.py:950]: Here the cotangent must have the same shape as the returns, I considered that it was on the user side to give the cotangents with the right shape.

File: frontend/catalyst/pennylane_extensions.py (hunk 1 above)
Comment (rmoyard) [reply frontend/catalyst/pennylane_extensions.py:962]: For the general question: our vjp is very different from Jax vjp https://jax.readthedocs.io/en/latest/_autosummary/jax.vjp.html where they return.

```
res, f_vjp = tuple(res, f_vjp)
```
Here res are unflatten, after that you need to use the function to get the vjps
```
vjps = f_vjp(cot)
```

File: frontend/catalyst/pennylane_extensions.py (hunk 4 above)
Comment (rmoyard) [reply frontend/catalyst/pennylane_extensions.py:884]: Yes that was the behaviour without any tree that I have kept, I will double check that.

File: frontend/catalyst/pennylane_extensions.py (hunk 3 above)
Comment (dime10) [reply frontend/catalyst/pennylane_extensions.py:950]: Right they should match, so normally I think the user would match the PyTree shape for the results and contangents.

Comment (dime10) [reply frontend/catalyst/pennylane_extensions.py:950]: There is also the issue of params which could be PyTrees

File: frontend/catalyst/pennylane_extensions.py (hunk 4 above)
Comment (josh146) [reply frontend/catalyst/pennylane_extensions.py:884]: Yep can confirm, the motivation was to match JAX (even if there were breaking changes!)

File: frontend/catalyst/pennylane_extensions.py (hunk 1 above)
Comment (dime10) [reply frontend/catalyst/pennylane_extensions.py:962]: That's true about the vjp, I was mainly thinking of the PyTree behaviour for inputs, outputs, tangents, cotangents, and gradients. Those should ideally all match JAX's version.

File: frontend/catalyst/pennylane_extensions.py (hunk 2 above)
Comment (rmoyard) [reply frontend/catalyst/pennylane_extensions.py:958]: The VJP should have the same shape as the parameters it is correct but the goal here was to return pytrees not to have them as arguments.

File: frontend/catalyst/pennylane_extensions.py (hunk 1 above)
Comment (dime10) [reply frontend/catalyst/pennylane_extensions.py:962]: About the vjp difference, I think we should still return a tuple of `(results, gradients)` just like for the jvp, because like you say we use the same function style for both.

File: frontend/catalyst/pennylane_extensions.py (hunk 3 above)
Comment (rmoyard) [reply frontend/catalyst/pennylane_extensions.py:950]: Yes here I did not consider the params as pytrees.

File: frontend/catalyst/pennylane_extensions.py (hunk 4 above)
Comment (rmoyard) [reply frontend/catalyst/pennylane_extensions.py:884]: Yes let me double check that, because I have rework all the test to make sure the pytrees are the same and they are.

File: frontend/catalyst/pennylane_extensions.py (hunk 2 above)
Comment (dime10) [reply frontend/catalyst/pennylane_extensions.py:958]: Did you plan on adding PyTree argument support in a separate PR? I think in this case tackling them together would be effective due to how the cotangent and gradient results are linked for example.

Comment (rmoyard) [reply frontend/catalyst/pennylane_extensions.py:958]: I can do that in another PR, but it was not part of this quarter goals. (Happy to add it to shorcut)

Comment (dime10) [reply frontend/catalyst/pennylane_extensions.py:958]: done in #549

File: frontend/catalyst/pennylane_extensions.py (hunk 3 above)
Comment (dime10) [reply frontend/catalyst/pennylane_extensions.py:950]: done in #549

File: frontend/catalyst/pennylane_extensions.py (hunk 4 above)
Comment (dime10) [reply frontend/catalyst/pennylane_extensions.py:884]: done in #508
